### ISSUE TITLE
Add Download as an approved verb

### ISSUE THREAD

**Issue by powercode** [PowerShell/PowerShell#21142]
### Summary of the new feature / enhancement

As a scriptwriter, I find it hard to name functions that download stuff. I struggle to come up with a verb to use, and neither _Copy,_ _Read_, _Save_ nor _Install_ seems to capture my intention. 

Verb |   Description
----   | -----------
Copy |   Copies a resource to another name or to another container
Read  |  Acquires information from a source
Save  |  Preserves data to avoid loss
Install |Places a resource in a location, and optionally initializes it

If we don't add the verb, we should at least provide guidance on what verb to use instead of _Download_ here:

 https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands

### Proposed technical implementation details (optional)

- Add a `Download` member to the `VerbsData` class in `src\System.Management.Automation\utils\Verbs.cs` 
- Add a description in `src\System.Management.Automation\resources\VerbDescriptionStrings.resx`

**Comment by mklement0** [comment]
I agree that at least the docs should be amended.

De facto, `Save` is the verb the `PackageManagement` and `PowerShellGet` modules as well as the latter's successor, `Microsoft.PowerShell.PSResourceGet`, use for their download cmdlets.
(The problem is sidestepped in the multi-purpose `Invoke-WebRequest` / `Invoke-RestMethod` cmdlets, where downloading is just _one_ of the purposes.)

**Comment by rhubarb-geek-nz** [comment]
Download is just a Get where you choose to put the result somewhere.

> Get
> Specifies an action that retrieves a resource. This verb is paired with Set.

**Comment by 237dmitry** [comment]
`Get` is similar to wget ))

**Comment by mklement0** [comment]
In the context of PowerShell, `Get` connotes _in-memory_ results, whereas `Save` connotes saving the results in a _persistent store_, typically a file. 
A hypothetical `Download` verb would conceptually _combine_ getting and saving, while implying a server source of some sort - as such, it may be too specific.
(Re introducing an HTTP provider that would allow `Get-Content` to target URLs, for instance: this has come up before: https://github.com/PowerShell/PowerShell/issues/8835)

**Comment by rhubarb-geek-nz** [comment]
> A hypothetical `Download` verb would conceptually _combine_ getting and saving, while implying a server source of some sort - as such, it may be too specific.

I suggest the existing verb Copy covers that scenario.

**Comment by kilasuit** [comment]
Personally I think this should be a docs update to the `Save` approved verb as this already does allow you to do this as does Copy

Also with the likes of both `Invoke-WebRequest` & `Invoke-RestMethod` having `-OutFile` & `Copy-Item` having the `-FromSession` & `-ToSession` parameters.
Perhaps the types of sessions that this works with can be extended by enabling other types of sessions like a WebRequest or SFTP/FTP and other types too,   

As such I'm inclined to think that a dedicated Verb for this isn't really needed that much.

Can you provide more examples of what sort of downloading you wanting to do here that Save or Copy would not work for?

**Comment by powercode** [comment]
Save has _Preserves data to avoid loss_ as the description, and I never found that to be a match for downloading a file.

But maybe amend the description of Save to include _download_ scenarios?

**Comment by rhubarb-geek-nz** [comment]
> Save has _Preserves data to avoid loss_ as the description, and I never found that to be a match for downloading a file.

I don't think it refers to moving data from one place to another. In a standard application menu, `File` → `Save` just saves the existing file version, you would need `File` → `Save As...` to do something equivalent to a download.

I would think the save verb is closer to a `git commit` than a `git clone`.

**Comment by jhoneill** [comment]
Just to stick my oar into the muddy waters.
Upload is **copy from local to remote** and download is **copy to local from remote**   (or Move if the original is not kept) 
I have written Copy-ToOneDrive and Copy-FromOneDrive as functions.   
One could argue "Sending" [to remote] and Downloading is "Receiving" or "Requesting" [from remote]

Adding download and upload verbs to do the same operation but in different directions is probably not the right thing to do. 

In the past "Load" was "transfer from permanent storage into memory"  (most senses of 'load' in English are 'put into', or 'that which been put into')  and its companion "Save" was "Transfer from memory to permanent storage".  ('Save' is keep, preserve etc)  Now applications typically have Open file, Close and Save file, Close and don't Save, Save without Closing. 
Instead of "Up-save" and "Down-save" when because destination is permanent, we're saying "Put there / Put here" with Upload and Download instead of "Keep here / keep there". 

PowerShell doesn't use "PUT" ; `SET` and `EXPORT` as well as `SAVE` do "Transfer to"  `Save` is used in some modules which primarily fetch and do something with a file for "Copy here without taking any action" and `Invoke-WebRequest` should probably have been `Request-Web`  but things are what they are. 

When talking about a file which will exist in an extra place after the operation, `Copy` is seems to me to the right verb

**Comment by rhubarb-geek-nz** [comment]
> `Invoke-WebRequest` should probably have been `Request-Web` but things are what they are.

I see a 'WebRequest' as being a noun, just like a 'Command', 'RestMethod' or 'Expression'. The verb Invoke indicates you want it performed synchronously rather than asynchronously (Start/Stop/Wait).

**Comment by ThomasNieto** [comment]
I've always thought Send and Receive were closest to Upload and Download. While Save does work due to precedent there is no paired verb for upload.

**Comment by powercode** [comment]
Well, one clear thing is that I'm one of many who are more or less confused by this.

From the comments here, I gather that `Copy` is most likely the semantically closest verb and that it's a bit unfortunate that `Save` has been used in the shipped modules.

It is obviously possible to give reasonable arguments for several options here.

I also note that a comparison can be made with `ConvertFrom/To`.
`CopyFrom-Web`, `CopyFrom-Ftp`, and `CopyFrom-OneDrive` would all be quite clear, wouldn't they?
Same with `CopyTo-*`.

Has that ship already sailed?

**Comment by rhubarb-geek-nz** [comment]
> I also note that a comparison can be made with `ConvertFrom/To`. `CopyFrom-Web`, `CopyFrom-Ftp`, and `CopyFrom-OneDrive` would all be quite clear, wouldn't they? Same with `CopyTo-*`.

That would would require a matrix of CopyFrom/CopyTo depending on the source and destination. If I am copying from Ftp to the Web do I use CopyFrom-Ftp or CopyTo-Web? It would be far better for Copy-Item that could with URL as source and destination than have all the different CopyFrom/CopyTo. I suggest we need someway for providers to be handling this complexity rather than multiple variations of the commands.

**Comment by sirtao** [comment]
If I may: wouldn't `Copy-RemoteItem` be the best solution?   
Down\Up-loading IS copying, and the noun would make it clear it's not a local action plus would not confuse autocomplete.
And the grammar, aside the protocol-specific parameters would\should stay (mostly?) the same.

**Comment by rhubarb-geek-nz** [comment]
The Get-Item/Copy-Item etc all happily work for the local filesystem and file sharing where remote drives are already mounted or you have credentials already established to do that.

The idea of "RemoteItem" sounds great when you have publicly available resources which (a) use standard REST and (b) require no credentials. Once you diverge from REST or require credentials it collapses very quickly under its own weight, it isn't just a matter of supporting authentication mechanisms, it is knowing which credentials to use and when. Then add AWS S3 buckets and Azure Blobs as remote items and it becomes a monster.

So either you restrict the scenarios so much that Invoke-WebRequest will already do what you want or else you have to have dynamic providers and authenticators to match each scenario.
